File: utils/src/main/java/ai/elimu/analytics/utils/LearningEventUtil.java
package ai.elimu.analytics.utils;

import android.content.Context;
import android.content.Intent;
import android.util.Log;

import java.util.stream.Collectors;

import ai.elimu.model.v2.enums.analytics.LearningEventType;
import ai.elimu.model.v2.gson.content.LetterGson;
import ai.elimu.model.v2.gson.content.LetterSoundCorrespondenceGson;
import ai.elimu.model.v2.gson.content.SoundGson;
import ai.elimu.model.v2.gson.content.StoryBookGson;
import ai.elimu.model.v2.gson.content.WordGson;

/**
 * A utility class that makes it easier for other apps to report learning events.
 */
public class LearningEventUtil {

    /**
     * @param letterGson The letter that the student is learning.
     * @param learningEventType The type of learning (i.e. the learning format) that is presented to the student in the application ({@code packageName}).
     * @param context Needed to fetch the {@code packageName} of the application where the learning event occurred.
     * @param analyticsApplicationId The package name of the analytics application that will receive the Intent and store the event.
     */
    public static void reportLetterLearningEvent(LetterGson letterGson, LearningEventType learningEventType, Context context, String analyticsApplicationId) {
        Log.i(LearningEventUtil.class.getName(),"reportLetterLearningEvent");

        Intent broadcastIntent = new Intent();
        broadcastIntent.setAction("ai.elimu.intent.action.LETTER_LEARNING_EVENT");
        broadcastIntent.putExtra("packageName", context.getPackageName());
        broadcastIntent.putExtra("letterId", letterGson.getId());
        broadcastIntent.putExtra("letterText", letterGson.getText());
        broadcastIntent.putExtra("learningEventType", learningEventType.toString());
        broadcastIntent.setPackage(analyticsApplicationId);
        context.sendBroadcast(broadcastIntent);
    }

    /**
     * @param letterSoundGson The letter sound that the student is learning.
     * @param context Needed to fetch the {@code packageName} of the application where the learning event occurred.
     * @param analyticsApplicationId The package name of the analytics application that will receive the Intent and store the event.
     */
    public static void reportLetterSoundLearningEvent(LetterSoundCorrespondenceGson letterSoundGson, Context context, String analyticsApplicationId) {
        Log.i(LearningEventUtil.class.getName(),"reportLetterSoundLearningEvent");

        Intent broadcastIntent = new Intent();
        broadcastIntent.setAction("ai.elimu.intent.action.LETTER_SOUND_LEARNING_EVENT");
        broadcastIntent.putExtra("packageName", context.getPackageName());
        broadcastIntent.putExtra("letterSoundId", letterSoundGson.getId());
        broadcastIntent.putExtra("letterSoundLetterTexts", letterSoundGson.getLetters().stream().map(LetterGson::getText).toArray());
        broadcastIntent.putExtra("letterSoundSoundValuesIpa", letterSoundGson.getSounds().stream().map(SoundGson::getValueIpa).toArray());
        broadcastIntent.setPackage(analyticsApplicationId);
        context.sendBroadcast(broadcastIntent);
    }

    /**
     * @param wordGson The word that the student is learning.
     * @param learningEventType The type of learning (i.e. the learning format) that is presented to the student in the application ({@code packageName}).
     * @param context Needed to fetch the {@code packageName} of the application where the learning event occurred.
     * @param analyticsApplicationId The package name of the analytics application that will receive the Intent and store the event.
     */
    public static void reportWordLearningEvent(WordGson wordGson, LearningEventType learningEventType, Context context, String analyticsApplicationId) {
        Log.i(LearningEventUtil.class.getName(),"reportWordLearningEvent");

        Intent broadcastIntent = new Intent();
        broadcastIntent.setAction("ai.elimu.intent.action.WORD_LEARNING_EVENT");
        broadcastIntent.putExtra("packageName", context.getPackageName());
        broadcastIntent.putExtra("wordId", wordGson.getId());
        broadcastIntent.putExtra("wordText", wordGson.getText());
        broadcastIntent.putExtra("learningEventType", learningEventType.toString());
        broadcastIntent.setPackage(analyticsApplicationId);
        context.sendBroadcast(broadcastIntent);
    }

    /**
     * @param storyBookGson The storybook that the student is learning from.
     * @param learningEventType The type of learning (i.e. the learning format) that is presented to the student in the application ({@code packageName}).
     * @param context Needed to fetch the {@code packageName} of the application where the learning event occurred.
     * @param analyticsApplicationId The package name of the analytics application that will receive the Intent and store the event.
     */
    public static void reportStoryBookLearningEvent(StoryBookGson storyBookGson, LearningEventType learningEventType, Context context, String analyticsApplicationId) {
        Log.i(LearningEventUtil.class.getName(),"reportStoryBookLearningEvent");

        Intent broadcastIntent = new Intent();
        broadcastIntent.setAction("ai.elimu.intent.action.STORYBOOK_LEARNING_EVENT");
        broadcastIntent.putExtra("packageName", context.getPackageName());
        broadcastIntent.putExtra("storyBookId", storyBookGson.getId());
        broadcastIntent.putExtra("storyBookTitle", storyBookGson.getTitle());
        broadcastIntent.putExtra("learningEventType", learningEventType.toString());
        broadcastIntent.setPackage(analyticsApplicationId);
        context.sendBroadcast(broadcastIntent);
    }
}

File: utils/src/main/java/ai/elimu/analytics/utils/LearningEventUtil.kt
package ai.elimu.analytics.utils

import ai.elimu.analytics.utils.receiver.ErrorResultReceiver
import ai.elimu.model.v2.gson.content.LetterGson
import ai.elimu.model.v2.gson.content.LetterSoundGson
import ai.elimu.model.v2.gson.content.NumberGson
import ai.elimu.model.v2.gson.content.SoundGson
import ai.elimu.model.v2.gson.content.StoryBookGson
import ai.elimu.model.v2.gson.content.VideoGson
import ai.elimu.model.v2.gson.content.WordGson
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.util.Log
import android.widget.Toast
import org.json.JSONObject
import java.util.stream.Collectors

/**
 * A utility class that makes it easier for other apps to report learning events to the receivers
 * in the `:app` module.
 */
object LearningEventUtil {

    const val BROADCAST_INTENT_ACTION_ANALYTICS = "ai.elimu.intent.action.ANALYTICS_EVENT"

    /**
     * @param letterSoundGson The letter-sound correspondence that the student is learning.
     * @param additionalData Any additional data related to the learning event, e.g. `{'is_letter_pressed':true}`
     * @param context Needed to fetch the `packageName` of the application where the event occurred.
     * @param analyticsApplicationId The package name of the analytics application that will receive the Intent and store the event.
     */
    fun reportLetterSoundLearningEvent(
        letterSoundGson: LetterSoundGson,
        additionalData: JSONObject? = null,
        context: Context,
        analyticsApplicationId: String
    ) {
        Log.i(this::class.simpleName, "reportLetterSoundLearningEvent")

        try {
            val broadcastIntent = Intent()
            broadcastIntent.setPackage(analyticsApplicationId)
            broadcastIntent.setAction("ai.elimu.intent.action.LETTER_SOUND_LEARNING_EVENT")
            broadcastIntent.putExtra(BundleKeys.KEY_PACKAGE_NAME, context.packageName)
            additionalData?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_ADDITIONAL_DATA, additionalData.toString())
            }

            broadcastIntent.putExtra(
                BundleKeys.KEY_LETTER_SOUND_LETTER_TEXTS,
                letterSoundGson.letters.stream().map { obj: LetterGson -> obj.text }
                    .collect(Collectors.toList()).toTypedArray()
            )

            val letterSoundSoundValuesIpa = letterSoundGson.sounds.stream().map {
                obj: SoundGson -> obj.valueIpa
            }.collect(Collectors.toList()).toTypedArray()
            broadcastIntent.putExtra(BundleKeys.KEY_LETTER_SOUND_SOUND_VALUES_IPA, letterSoundSoundValuesIpa)

            letterSoundGson.id?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_LETTER_SOUND_ID, letterSoundGson.id)
            }

            context.sendOrderedBroadcast(broadcastIntent, null, ErrorResultReceiver(), null, Activity.RESULT_OK, null, null)
        } catch (e: Exception) {
            Log.e(this::class.simpleName, "Error during Intent preparation", e)
            Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
        }
    }

    /**
     * @param wordGson The word that the student is learning.
     * @param additionalData Any additional data related to the learning event, e.g. `{'is_word_pressed':true}`
     * @param context Needed to fetch the `packageName` of the application where the event occurred.
     * @param analyticsApplicationId The package name of the analytics application that will receive the Intent and store the event.
     */
    fun reportWordLearningEvent(
        wordGson: WordGson,
        additionalData: JSONObject? = null,
        context: Context,
        analyticsApplicationId: String
    ) {
        Log.i(this::class.simpleName, "reportWordLearningEvent")

        try {
            val broadcastIntent = Intent()
            broadcastIntent.setPackage(analyticsApplicationId)
            broadcastIntent.setAction("ai.elimu.intent.action.WORD_LEARNING_EVENT")
            broadcastIntent.putExtra(BundleKeys.KEY_PACKAGE_NAME, context.packageName)
            additionalData?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_ADDITIONAL_DATA, additionalData.toString())
            }
            broadcastIntent.putExtra(BundleKeys.KEY_WORD_TEXT, wordGson.text)
            wordGson.id?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_WORD_ID, wordGson.id)
            }

            context.sendOrderedBroadcast(broadcastIntent, null, ErrorResultReceiver(), null, Activity.RESULT_OK, null, null)
        } catch (e: Exception) {
            Log.e(this::class.simpleName, "Error during Intent preparation", e)
            Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
        }
    }

    /**
     * @param storyBookGson The storybook that the student is learning from.
     * @param additionalData Any additional data related to the learning event, e.g. `{'time_spent_per_chapter_ms':[7500,12900,34422,5023]}`
     * @param context Needed to fetch the `packageName` of the application where the event occurred.
     * @param analyticsApplicationId The package name of the analytics application that will receive the Intent and store the event.
     */
    fun reportStoryBookLearningEvent(
        storyBookGson: StoryBookGson,
        additionalData: JSONObject? = null,
        context: Context,
        analyticsApplicationId: String
    ) {
        Log.i(this::class.simpleName, "reportStoryBookLearningEvent")

        try {
            val broadcastIntent = Intent()
            broadcastIntent.setPackage(analyticsApplicationId)
            broadcastIntent.setAction("ai.elimu.intent.action.STORYBOOK_LEARNING_EVENT")
            broadcastIntent.putExtra(BundleKeys.KEY_PACKAGE_NAME, context.packageName)
            additionalData?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_ADDITIONAL_DATA, additionalData.toString())
            }
            broadcastIntent.putExtra(BundleKeys.KEY_STORYBOOK_TITLE, storyBookGson.title)
            storyBookGson.id?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_STORYBOOK_ID, storyBookGson.id)
            }

            context.sendOrderedBroadcast(broadcastIntent, null, ErrorResultReceiver(), null, Activity.RESULT_OK, null, null)
        } catch (e: Exception) {
            Log.e(this::class.simpleName, "Error during Intent preparation", e)
            Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
        }
    }

    /**
     * @param videoGson The video that the student is learning from.
     * @param additionalData Any additional data related to the learning event, e.g. `{'is_video_paused':true}`
     * @param context Needed to fetch the `packageName` of the application where the event occurred.
     * @param analyticsApplicationId The package name of the analytics application that will receive the Intent and store the event.
     */
    fun reportVideoLearningEvent(
        videoGson: VideoGson,
        additionalData: JSONObject? = null,
        context: Context,
        analyticsApplicationId: String
    ) {
        Log.i(this::class.simpleName, "reportVideoLearningEvent")

        try {
            val broadcastIntent = Intent()
            broadcastIntent.setPackage(analyticsApplicationId)
            broadcastIntent.setAction("ai.elimu.intent.action.VIDEO_LEARNING_EVENT")
            broadcastIntent.putExtra(BundleKeys.KEY_PACKAGE_NAME, context.packageName)
            additionalData?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_ADDITIONAL_DATA, additionalData.toString())
            }
            broadcastIntent.putExtra(BundleKeys.KEY_VIDEO_TITLE, videoGson.title)
            videoGson.id?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_VIDEO_ID, videoGson.id)
            }

            context.sendOrderedBroadcast(broadcastIntent, null, ErrorResultReceiver(), null, Activity.RESULT_OK, null, null)
        } catch (e: Exception) {
            Log.e(this::class.simpleName, "Error during Intent preparation", e)
            Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
        }
    }

    /**
     * @param numberGson The number that the student is learning.
     * @param additionalData Any additional data related to the learning event, e.g. `{'is_number_pressed':true}`
     * @param context Needed to fetch the `packageName` of the application where the event occurred.
     * @param analyticsApplicationId The package name of the analytics application that will receive the Intent and store the event.
     */
    fun reportNumberLearningEvent(
        numberGson: NumberGson,
        additionalData: JSONObject? = null,
        context: Context,
        analyticsApplicationId: String
    ) {
        Log.i(this::class.simpleName, "reportNumberLearningEvent")

        try {
            val broadcastIntent = Intent()
            broadcastIntent.setPackage(analyticsApplicationId)
            broadcastIntent.setAction("ai.elimu.intent.action.NUMBER_LEARNING_EVENT")
            broadcastIntent.putExtra(BundleKeys.KEY_PACKAGE_NAME, context.packageName)
            additionalData?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_ADDITIONAL_DATA, additionalData.toString())
            }
            broadcastIntent.putExtra(BundleKeys.KEY_NUMBER_VALUE, numberGson.value)
            numberGson.symbol?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_NUMBER_SYMBOL, numberGson.symbol)
            }
            numberGson.id?.let {
                broadcastIntent.putExtra(BundleKeys.KEY_NUMBER_ID, numberGson.id)
            }

            context.sendOrderedBroadcast(broadcastIntent, null, ErrorResultReceiver(), null, Activity.RESULT_OK, null, null)
        } catch (e: Exception) {
            Log.e(this::class.simpleName, "Error during Intent preparation", e)
            Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
        }
    }
}
